kml/minzoom_quadtree.hpp

```diff
@@ -0,0 +1,177 @@
+// system that uses the full range of uint32_t inside the bounding
+// (squared) box.
+template <typename Value, typename Less>
+class MinZoomQuadtree
+{
+public:
+  MinZoomQuadtree(Less const & less) : m_less{less} {}
+
+  void Add(m2::PointD const & point, Value const & value) { m_quadtree.emplace_back(point, value); }
+  void Add(m2::PointD const & point, Value && value) { m_quadtree.emplace_back(point, std::move(value)); }
+
+  void Clear() { m_quadtree.clear(); }
+
+  template <typename F>
+  void SetMinZoom(double countPerTile, int maxZoom, F setMinZoom /* void (Value & value, int minZoom) */)
+  {
+    CHECK_GREATER(countPerTile, 0.0, ());
+    CHECK_LESS_OR_EQUAL(maxZoom, scales::GetUpperStyleScale(), ());
+    CHECK_GREATER_OR_EQUAL(maxZoom, 1, ());
+
+    if (m_quadtree.empty())
+      return;
+
+    m2::RectD bbox;
+    for (auto const & elem : m_quadtree)
+      bbox.Add(elem.m_point);
+
+    if (bbox.IsEmptyInterior())
+      return;
+
+    auto spacing = std::max(bbox.SizeX() / mercator::Bounds::kRangeX,
+                            bbox.SizeY() / mercator::Bounds::kRangeY);
+    spacing *= std::sqrt(countPerTile);
+
+    int baseZoom;
+    // `spacing` is a characteristic interval between values on the map as if they spaced
+    // uniformly across the map providing required density (i.e. "count per tile area"
+    // on highest zoom level).
+    // `spacing` value decomposed into a normalized fraction and an integral power of two.
+    // Inverse of normalized fraction `scale` (in [0.5, 1)) used to slightly enlarge bounding box.
+    // Integral power of two `baseZoom` is biased exponent of inverse value.
+    // It is corrected to be unbiased (by subtracting 1) and negated to match the common
+    // sense of "zoom level" (currently zero-based). Advance increment in the very top
+    // level call of `traverse()` makes it one-based so it effectively became `2 - baseZoom`,
+    // which have exactly the same meaning as another "zoomLevel"s in the project.
+    // Scaling (enlarging) the size of bbox by `1 / scale` is required so that one can
+    // set arbitrary values for `countPerTile`, not only a powers of four.
+    auto const scale = std::frexp(spacing, &baseZoom);
+    baseZoom = 1 - baseZoom;
+
+    if (baseZoom >= maxZoom)
+      return;
+
+    auto const size = std::max(bbox.SizeX(), bbox.SizeY()) / scale;
+    bbox.SetSizes(size, size);
+
+    // Calculate coordinate along z-order curve.
+    auto const [minX, minY] = bbox.LeftBottom();
+    auto const gridScale = std::numeric_limits<uint32_t>::max() / size;
+    for (auto & elem : m_quadtree)
+    {
+      auto const [x, y] = elem.m_point;
+      auto const gridX = static_cast<uint32_t>(std::trunc((x - minX) * gridScale));
+      auto const gridY = static_cast<uint32_t>(std::trunc((y - minY) * gridScale));
+      // Sacrifice one of 32 possible levels for simplicity of traversing.
+      elem.m_zCurveCoord = bits::BitwiseMerge(gridX, gridY) >> 2;
```

Comment:
Won't it be simpler to leave this one as is (without shifting right) and instead of
```
treeLevelMask = 0b11 <<62;
...
treeLevelMask >>= 2;
```

do something along the lines of
```
treeLevelMask = 0b11 << depth;
quadIncrement = 1 << depth;
traverse(depth - 1);
```

Reply:
Then it should be `(64 - (depth + 1) * 2)` in rhs of binary shift left operations. Because root's bits are at msb side of a whole bitmask. Looks cryptic, isn't it?

Reply:
I mean it for zero-based "depth".

Reply:
64 - (depth + 1) * 2
is
62 - 2 * depth

or just 2*x with x = 31 - depth, so start with 31 instead of 0

I'm not sure yet whether it is better because I'm currently having trouble understanding the baseZoom concept, in particular the fact that traverse starts at top-level but zoom starts at baseZoom

Reply:
It is common, that depth is 0 (or rarely 1) at root level and grows for deeper levels. It hardly that introducing `depth` as explicit parameter makes things much clearer, because it have its own complexity cost, that outweighs profit at my mind. In the case `treeLevelMask` must be calculated, not just "adjusted" by shifting right, which is quite clear operation, when deal with bitmasks.

Reply:
My problem here is that zoom is a parameter to recursive call that is increased by 1 when moving to children and whose large value is used as the exit condition

So from reading ```traverse``` alone I would seem it is the "zoom level of the current level of the tree", which starts at 0. But in fact it is something else (and starts with baseZoom).

So, what I think is hard to understand is that ```traverse``` starts at the root level with a non-root depth-like parameter.